### PR TITLE
Common module

### PR DESCRIPTION
**Pull request by pikinier20**

### CODE REVIEW

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 1]
```diff
@@ -1,14 +1,87 @@
 package org.virtuslab.inkuire.engine.model
+import java.nio.file.{Files, Path}
+
+import com.google.gson.Gson
+import org.virtuslab.inkuire.model.{SBound, SDAnnotation, SDClass, SDEnum, SDEnumEntry, SDFunction, SDInterface, SDModule, SDObject, SNullable, SOtherParameter, SPrimitiveJavaType, STypeConstructor, SVoid}
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:5]: how about just `import org.virtuslab.inkuire.model._`?

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 2]
```diff
@@ -1,14 +1,87 @@
...
+  )
+
+  def read(path: Path): InkuireDb = {
+    val source = Files.readString(path)
+    parseSource(source)
+  }
+
+  def read(text: String): InkuireDb = parseSource(text)
+
+  private def parseSource(source: String) : InkuireDb = {
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:61]: some extra space there (before `:`)

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 3]
```diff
@@ -1,14 +1,87 @@
...
+    val module = new CustomGsonFactory().getInstance().fromJson(source, classOf[SDModule])
+    mapModule(module)
+  }
+
+  def mapModule(module: SDModule): InkuireDb = {
+    val globalFunctions = module.getPackages.flatMap(p => p.getFunctions).map { f =>
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:67]: `flatMap(_.getFunctions)` should work

Comment (KacperFKorban) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:67]: `.map(generateFunctionSignature)` as well

Explicit parameters are only needed in with some weird conversions and stuff like that.

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 4]
```diff
@@ -1,14 +1,87 @@
...
+      generateFunctionSignature(f)
+    }
+
+    val methods = module.getPackages.flatMap(p => p.getClasslikes).flatMap(c =>
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:71]: `_.getClasslikes`

maybe for second `flatMap` use curly braces `{}`

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 5]
```diff
@@ -1,14 +1,87 @@
...
+      c match{
+        case i: SDInterface => i.getFunctions
+        case e: SDEnum => e.getFunctions ++ e.getEntries.flatMap(ee => ee.getFunctions)
+        case o: SDObject => o.getFunctions
+        case c: SDClass => c.getFunctions
+        case a: SDAnnotation => a.getFunctions
+        case _ => List.empty
+    }
+    ).map {
+      f => generateFunctionSignature(f)
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:81]: `.map(generateFunctionSignature)`

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 6]
```diff
@@ -1,14 +1,87 @@
...
+  private def receiver(f: SDFunction) = {
+    val receiver = f.getReceiver
+    val className = f.getDri.getClassName
+    if(receiver == null) {
+      if (className == null) None
+      else Some(className.concreteType)
+    } else Some(receiver.getName.concreteType)
+  }
+
+  private def generateFunctionSignature(f: SDFunction) = ExternalSignature(
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:43]: Is nullability mapped to Scala model?

If it isn't then it should probably be done in a separate task.

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 7]
```diff
@@ -1,14 +1,87 @@
...
+
+    val methods = module.getPackages.flatMap(p => p.getClasslikes).flatMap(c =>
+      c match{
+        case i: SDInterface => i.getFunctions
+        case e: SDEnum => e.getFunctions ++ e.getEntries.flatMap(ee => ee.getFunctions)
+        case o: SDObject => o.getFunctions
+        case c: SDClass => c.getFunctions
+        case a: SDAnnotation => a.getFunctions
+        case _ => List.empty
+    }
```
Comment (KacperFKorban) [inline inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:79]: I think that this should be 2 spaces to the right

File: .idea/compiler.xml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="CompilerConfiguration">
-    <bytecodeTargetLevel target="11" />
+    <bytecodeTargetLevel target="1.8" />
```
[the remark this answers is not in the document]
Comment (pikinier20) [reply .idea/compiler.xml]: changed

File: .idea/scala_compiler.xml [hunk 9]
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="ScalaCompilerConfiguration">
-    <profile name="Gradle 1" modules="Inkuire.inkuire-engine.main,Inkuire.inkuire-engine.test">
+    <profile name="Gradle 2" modules="Inkuire.inkuire-engine.main,Inkuire.inkuire-engine.test">
       <option name="deprecationWarnings" value="true" />
       <option name="uncheckedWarnings" value="true" />
       <parameters>
-        <parameter value="-target:jvm-11" />
+        <parameter value="-target:jvm-1.8" />
```
[the remark this answers is not in the document]
Comment (pikinier20) [reply .idea/scala_compiler.xml]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala [hunk 10]
```diff
@@ -1,14 +1,87 @@
...
   types: Map[Type, Set[Type]]
 )
 
 object InkuireDb {
-  def read(path: String): InkuireDb = {
-    // TODO read db from given path
-    new InkuireDb(Seq.empty, Map.empty)
+  implicit def listAsScala[T](list: java.util.List[T]) : Iterable[T] = list.asScala
+
+  @scala.annotation.tailrec
```
[the remark this answers is not in the document]
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 1 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 2 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 3 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 4 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 5 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 6 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala:45]: changed

File: inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala (hunk 7 above)
Comment (pikinier20) [reply inkuire-engine/src/main/scala/org/virtuslab/inkuire/engine/model/InkuireDb.scala]: changed

File: inkuire-engine/src/test/resources/testData/dokka.json [hunk 11]
```diff
@@ -0,0 +1,1441 @@
+{
```
Comment (KacperFKorban) [inline inkuire-engine/src/test/resources/testData/dokka.json:1]: rm file

**Pull request merged**